src/bluetooth_linux.c

```diff
@@ -0,0 +1,466 @@
+					if (!strcmp(uuid, FIDO_SERVICE_UUID))
+						service_found = true;
+				}
+			} else {
+				sd_bus_message_skip(reply, "v");
+			}
+		}
+		sd_bus_message_exit_container(reply);
+	}
+	sd_bus_message_exit_container(reply);
+	if (matches && service_found) {
+		newdev->paths.service = strdup(path);
+	}
+}
+
+static void collect_device_chars(void *data, const char *path, sd_bus_message *reply)
+{
+	struct ble *newdev = (struct ble *)data;
+	const char *iface;
+	if (sd_bus_message_read_basic(reply, 's', &iface) >= 0) {
+		if (!strcmp(iface, DBUS_SERVICE_IFACE))
+			found_gatt_service(newdev, path, reply);
+		else if (!strcmp(iface, DBUS_CHAR_IFACE))
+			found_gatt_characteristic(newdev, path, reply);
+		else
+			sd_bus_message_skip(reply, "a{sv}");
+	}
+}
+
+static void iterate_over_all_objs(sd_bus_message *reply,
+				  void (*new_dbus_interface)(void *,
+				  const char *,
+				  sd_bus_message *), void *data)
+{
+	if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{oa{sa{sv}}}") <= 0)
+		return;
+
+	while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "oa{sa{sv}}")) {
+		const char *ifacepath = NULL;
+		if (sd_bus_message_read_basic(reply, 'o', &ifacepath) <= 0)
+			return;
+
+		if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{sa{sv}}") < 0)
+			return;
+		while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "sa{sv}")) {
+			new_dbus_interface(data, ifacepath, reply);
+			sd_bus_message_exit_container(reply);
+		}
+		sd_bus_message_exit_container(reply);
+		sd_bus_message_exit_container(reply);
+	}
+}
+
+void *
+fido_bluetooth_open(const char *path)
+{
+	struct ble *newdev;
+	sd_bus_message *reply = NULL;
+	int ret;
+	if (!fido_is_bluetooth(path))
+		return NULL;
+
+	path += strlen(FIDO_BLUETOOTH_PREFIX);
+
+	newdev = calloc(1, sizeof(*newdev));
+	if (!newdev)
+		return NULL;
+
+	newdev->paths.dev = strdup(path);
+	if (!newdev->paths.dev)
+		goto out;
+
+	if (0 > sd_bus_default_system(&newdev->bus))
+		goto out;
+
+	if (0 > sd_bus_call_method(newdev->bus, "org.bluez",
+		path, "org.freedesktop.DBus.Properties", "GetAll", NULL, &reply,
+		"s", DBUS_DEV_IFACE))
+		goto out;
+
+	if (!ble_fido_is_useable_device(DBUS_DEV_IFACE, reply, false, NULL))
+		goto out;
+
+	sd_bus_message_unref(reply);
+	reply = NULL;
+	ret = sd_bus_call_method(newdev->bus, "org.bluez", "/", "org.freedesktop.DBus.ObjectManager", "GetManagedObjects", NULL, &reply, "");
+	if (ret <= 0)
+		goto out;
+
+	sd_bus_message_rewind(reply, 1);
+	iterate_over_all_objs(reply, collect_device_chars, newdev);
+
+	sd_bus_message_unref(reply);
+	reply = NULL;
+
+	if (newdev->paths.status &&
+	    newdev->paths.control_point &&
+	    newdev->paths.control_point_length &&
+	    newdev->paths.service_revision) {
+		uint8_t cp_len[2];
+		uint8_t revision;
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.control_point_length,
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
```

Comment:
Nit: When breaking the line for a conditional, we typically align with the starting parenthesis.

```suggestion
		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.control_point_length,
		    DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
```